### DISCUSSION TITLE
Difference between /cppad_ipopt and /include/cppad/ipopt

### DISCUSSION THREAD

**Discussion by juanmanzanero** [coin-or/CppAD#130] [Q&A] [answered]
Hello! I frequently use Ipopt along with CppAD to solve optimal control problems, and everything works super good.

I just have a small question, what is the difference between /cppad_ipopt and /include/cppad/ipopt? Which one do you recommend to use? I am using the latter but I am curious to learn the differences. To me it seems that cppad_ipopt is more advanced (e.g. allows sparse Jacobian only to some equations), but just to confirm.

Thank you in advance.

Cheers!

Juan

**Comment by bradbell** [comment, marked as answer]
The following routine was deprecated on 2012-11-28
https://coin-or.github.io/CppAD/doc/cppad_ipopt_nlp.htm

You should instead use
https://coin-or.github.io/CppAD/doc/ipopt_solve.htm

There are other interfaces to ipopt with CppAD computing the derivatives; see
https://bradbell.github.io/cppad_mixed/doc/cppad_mixed.htm
This is more complitated to use.

If you use python, see
https://bradbell.github.io/cppad_py/doc/xsrst/mixed_fix_likelihood_xam_py.html

**Reply by juanmanzanero** [reply]
Thank you very much!
